### PR TITLE
add suggested CLI-only paths

### PR DESCRIPTION
**Pull request by khorne3**
~~Addresses comments from https://github.com/semgrep/semgrep-docs/pull/1344 but also separates out OSS vs Pro usage (these are currently tabs, but I am open to removing the tabs or splitting the doc).~~

There are two docs: one for Pro engine users and one for OSS users. I included what I thought might be a complete user journey for someone working solely in the CLI (except where absolutely necessary, such as signing in to SCP).

### Please ensure

- [x] A technical writer reviews the content or PR
- [x] This change has no security implications or else you have pinged the security team

### CODE REVIEW

**Comment by s-santillan** [comment]
Hi Katie! I see that you componentized the install instructions. Thank you!

There is this additional bullet from the original PR:

> One of the hardest things about the CLI workflow is dealing with the **OSS vs Pro workflows.** To help clarify (for us as well) the differences, maybe it's worth it to also include OSS workflows in the PR, then we can decide if we want OSS and Pro to be in 1 whole doc, or if we can separate them. We can include 1 file in 2 different sidebars, so this "CLI workflow doc" you started can be in both our Pro and OSS sidebars.

> Something that might help you write: https://semgrepinc.slack.com/archives/C8HRFBPSQ/p1699643585940479 - a discussion about the intentions behind the two commands semgrep ci and semgrep scan.

> There's also this somewhat unmaintained doc: https://semgrep.dev/docs/running-rules/ that you can check out.


Some additional context: This doc in particular is under "Semgrep Pro", which implicitly recommends a logged-in `semgrep ci`. However, there are also times when you want a `semgrep scan` (which we "tend" to associate with Semgrep OSS). That's why I was suggesting we put together both Semgrep Pro and OSS workflows and see how it looks in the grander scheme of things.

File: docs/getting-started/cli.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+semgrep ci --sarif --output=semgrep.sarif
+```
+
+To export the results to a JSON file:
+
+```console
+semgrep ci --json --output=semgrep.json
+```
+
+### Scan your project with a specific ruleset
```
Comment (s-santillan) [inline docs/getting-started/cli.md:65]: We actually do not want people to use `semgrep ci --config auto`:

https://semgrepinc.slack.com/archives/C8HRFBPSQ/p1699643585940479

**Comment by s-santillan** [comment]
There are several cases where we'd use `semgrep scan` for Semgrep Pro - the one I can think of is writing a **custom rule using the pro engine**. Can you do some digging into that? 

Another thing that may help is https://semgrep.dev/docs/kb/semgrep-ci/ci-vs-cli/

You can take some time to write this doc, a big part of it is contextualizing why the user would prefer one command over the other. That is, expect this doc to be a little opinionated towards best practices given some goal the user wants to achieve. They can use this doc as a jump start to: testing Semgrep overall, writing a rule, comparing speed & number of findings between our engines, debugging something, etc.

**Comment by s-santillan** [comment]
Last question - I actually cannot decide if we want one doc, or two separate ones (given that both semgrep scan and semgrep ci may have some use case overlap for both Pro and OSS). Do you have a preference?

**Comment by khorne3** [comment]
Two docs? I think there's a lot of ground to cover and from what it appears, quite a bit of change so it might be easier to manage/maintain if there are two docs even if there are overlaps

**Comment by s-santillan** [comment]
Sure, let's do it. And feel free to take some time on this (not necessarily on testing everything, but just accumulating interesting use cases). I don't think we're in a rush, and this has potential to be a really popular doc - see slack message.

File: docs/getting-started/cli-oss.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+To export the results to a JSON file:
+
+```console
+semgrep scan --json --output=semgrep.json
+```
+
+### Scan your project with a specific ruleset
+
+You can scan your project using `--config auto` to automatically obtain rules tailored to your project:
```
Comment (s-santillan) [inline docs/getting-started/cli-oss.md:53]:
```suggestion
You can scan your project using `--config auto` to run Semgrep with rules that apply to your to your codebase:
```
Let's stick to the ff definitions:

codebase - more general term ("broader term"); anything scanned by semgrep; does not need to be a git repo 
repository - same definition as ... repository (git is implied)
project - any repos onboarded to Semgrep Cloud Platform, as that is the term used in the UI.

File: docs/getting-started/cli-oss.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+```console
+semgrep scan --config auto .
+```
+
+:::info
+When using `--config auto`, Semgrep uses your project URL to log into the Semgrep registry.
```
Comment (s-santillan) [inline docs/getting-started/cli-oss.md:60]: In old versions of Semgrep, we would output text written like this (sounds weird to me) when starting a scan, so that users are aware their scan has some kind of identifier sent to the registry, even if it's pseudonymous to us (the Semgrep team).

**I suggest rewording this to say that Semgrep collects pseudonymous metrics when you use the Registry, because there's an implication here on a "Project URL" that may not exist on non-git codebases.**

File: docs/getting-started/cli-oss.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+## Scan your project
+
+# Semgrep OSS
+
+Navigate to the root of your repository, and run your first scan. The specific command you use depends on how you want to view the results.
+
+To view the results in the CLI:
+
+```console
+semgrep scan
```
Comment (s-santillan) [inline docs/getting-started/cli-oss.md:36]: If you don't specify a config I'm pretty sure this does nothing?

File: docs/getting-started/cli-oss.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+## Set up Semgrep
+
+<Install />
+
+## Scan your project
+
+# Semgrep OSS
+
+Navigate to the root of your repository, and run your first scan. The specific command you use depends on how you want to view the results.
```
Comment (s-santillan) [inline docs/getting-started/cli-oss.md:31]:
```suggestion
Navigate to the root of your codebase, and run your first scan. The specific command you use depends on how you want to view the results.
```
(See comment further down on codebase vs repo vs project)

File: docs/getting-started/cli.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+You can scan your project using `--config auto` to automatically obtain rules tailored to your project:
+
+```console
+semgrep scan --config auto .
```
Comment (s-santillan) [inline docs/getting-started/cli.md:64]: To ``.` or not to `.` - that is the question.

File: docs/getting-started/cli.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+## Scan using specific Semgrep Products
+
+When you run `semgrep ci`, you're scanning your project with Semgrep Code, Semgrep Supply Chain, and Semgrep Secrets. To scan your project with just one product, run:
+
+```console
+# scan with Semgrep Code
+semgrep ci --code
+
+# scan with Semgrep Supply Chain
+semgrep ci --supply chain
```
Comment (s-santillan) [inline docs/getting-started/cli.md:116]: Is there a dash on supply chain?

**Comment by s-santillan** [comment]
If you'd rather we cut this into several PRs, we can start with the commands you've outlined, then in a separate PR, add the following:

affects perf:
- `semgrep scan -j`

part of rule-writing workflow:
- `semgrep scan --test`
- `semgrep publish`

For this PR, because we included SARIF and JSON, let's include `--text` (for plaintext export).

File: docs/getting-started/cli-oss.md (hunk 4 above)
Comment (khorne3) [reply docs/getting-started/cli-oss.md:36]: It does (IDK if this is unintended behavior?)

image

Comment (s-santillan) [reply docs/getting-started/cli-oss.md:36]: This is new!! (Previously it would tell you to specify a config, and recommend config auto). Does the scan output say with what config it's scanning with?

Comment (khorne3) [reply docs/getting-started/cli-oss.md:36]: It does not, but the results are identical to `semgrep scan --config auto`

**Comment by khorne3** [comment]
@armchairlinguist @sinat101 @jkinsfather Whenever you have a moment, do you mind taking a look at this PR?

Thank you!

File: docs/getting-started/cli-oss.md (hunk 4 above)
Comment (s-santillan) [reply docs/getting-started/cli-oss.md:36]: :| ...we used to tell the user what rulesets are running regardless of log level. i'll make a ticket for the UX. thanks for looking into this.

File: docs/getting-started/cli.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
+---
+slug: cli
+title: CLI-only workflow (Pro)
```
Comment (s-santillan) [inline docs/getting-started/cli.md:3]:
```suggestion
title: Local CLI scans
```

File: docs/getting-started/cli.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+tags:
+  - quickstart
+  - cli
+---
+
+import MoreHelp from "/src/components/MoreHelp";
+import Install from "/src/components/procedure/_install-cli.mdx";
+import Login from "/src/components/procedure/_login-activate.mdx";
+
+# CLI-only workflow with Semgrep's Pro Engine
```
Comment (s-santillan) [inline docs/getting-started/cli.md:15]:
```suggestion
# Local scans with Semgrep Pro Engine
```
I appreciate the callout "with Semgrep Pro Engine" and happy we provide them with the OSS flag.

File: docs/getting-started/cli.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+Learn how to set up Semgrep, scan your project for security issues using Semgrep's Pro Engine, and view your findings in the CLI.
+
+:::info
+For scans in your local development environment, `semgrep scan` is the preferred method to scan with Semgrep, especially for writing and testing custom rules.
```
Comment (s-santillan) [inline docs/getting-started/cli.md:20]: Semgrep provides two commands to scan with in your CLI:

- `semgrep scan` - This is the recommended command for scanning local codebases, as well as writing and testing custom rules.
- `semgrep ci` - This is the recommended command if you are scanning repositories with Semgrep as part of an organization with custom rules and policies. `semgrep ci` fetches your organization's scan configurations from Semgrep Cloud Platform.
```suggestion
For scans in your local development environment, `semgrep scan` is the preferred method to scan with Semgrep, especially for writing and testing custom rules.
```

File: docs/getting-started/cli.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+:::
+
+## Prerequisites
+
+Before proceeding:
+
+* See [Prerequisites](/prerequisites/) to ensure that your machine meets Semgrep's requirements.
+* Ensure that you have and are logged into your [Semgrep Account](https://semgrep.dev/login).
```
Comment (s-santillan) [inline docs/getting-started/cli.md:28]:
```suggestion
* For scans using `semgrep ci`:
    * Ensure that you have and are logged into your [Semgrep Account](https://semgrep.dev/login).
```

File: docs/getting-started/cli.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+* Ensure that you've enabled the **Pro Engine** <i class="fa-solid fa-toggle-large-on"></i> toggle on Semgrep Cloud Platform's [Settings](https://semgrep.dev/orgs/-/settings) page.
```
Comment (s-santillan) [inline docs/getting-started/cli.md:29]:
```suggestion
    * Ensure that you've enabled the **Pro Engine** <i class="fa-solid fa-toggle-large-on"></i> toggle on Semgrep Cloud Platform's [Settings](https://semgrep.dev/orgs/-/settings) page.
```
Indentation

File: docs/getting-started/cli.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+## Set up Semgrep
+
+<Install />
+
+## Log into your Semgrep account
```
Comment (s-santillan) [inline docs/getting-started/cli.md:35]:
```suggestion
## Log in to your Semgrep account
```

File: docs/getting-started/cli-oss.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
+---
+slug: cli
+title: CLI-only workflow (OSS)
```
Comment (s-santillan) [inline docs/getting-started/cli-oss.md:3]:
```suggestion
title: Local CLI scans
```

**Review by s-santillan — APPROVED** [review]
Please don't forget to check out the last set of notes, and thanks for putting up with them! I'll fix the sidebars entries once it's merged.

File: docs/getting-started/cli-oss.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+### Scan your project with a specific ruleset
+
+You can scan your project using `--config auto` to run Semgrep with rules that apply to your codebase:
+
+```console
+semgrep scan --config auto
+```
+
+:::info
+Semgrep collects pseudonymous metrics when you use rules from the Registry.
```
Comment (armchairlinguist) [inline docs/getting-started/cli-oss.md:66]:
```suggestion
Semgrep collects pseudonymous metrics when you use rules from the Registry. You can turn this off with `--metrics=off`.
```

File: docs/getting-started/cli-oss.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+## Improve performance for large codebases
+
+You can set the number of subprocesses Semgrep uses to run checks in parallel:
+
+```console
+semgrep scan -j NUMBER_OF_SUBPROCESSES
+```
+
+By default, Semgrep defaults to the number of cores detected on the system on which Semgrep is running. For additional information, see [Parallelization](/kb/semgrep-code/scan-engine-kill)
```
Comment (armchairlinguist) [inline docs/getting-started/cli-oss.md:99]:
```suggestion
By default, Semgrep uses a number of jobs equivalent to the number of cores detected on the system. For additional information, see [Parallelization](/kb/semgrep-code/scan-engine-kill)
```
Most of the engine kill stuff in the link is honestly mainly aimed at Pro users, so not sure how useful it is to link to from here.

File: docs/getting-started/cli.md [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+
+<Login />
+
+You must log in to your Semgrep account before proceeding. Otherwise, you can't use `semgrep ci`.
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:39]: I thought we weren't going to say this just yet? @s-santillan

File: docs/getting-started/cli.md [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+### Scan your project with a specific ruleset
+
+You can scan your project using `--config auto` to automatically obtain rules tailored to your project:
+
+```console
+semgrep scan --config auto
+```
+
+:::info
+Semgrep collects pseudonymous metrics when you use rules from the Registry.
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:78]: Same change as my suggestion in the other doc - we should include how to turn metrics off

File: docs/getting-started/cli.md [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+```console
+semgrep ci --json --output=semgrep.json
+```
+
+### Scan your project with a specific ruleset
+
+You can scan your project using `--config auto` to automatically obtain rules tailored to your project:
+
+```console
+semgrep scan --config auto
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:74]:
```suggestion
semgrep scan --config auto --pro
```
If you're logged in but you run with scan, you need to specify Pro or it runs with OSS. Only `semgrep ci` can fetch whether you have the toggle enabled.

File: docs/getting-started/cli.md [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+To share your rules by adding them to the Semgrep Registry:
+
+```console
+semgrep publish <path/to/rules>
+```
+
+### Scan without sending results to Semgrep
+
+To scan your project **without** sending data to Semgrep, use:
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:113]:
```suggestion
To scan your project using the configuration you've set up in Semgrep Cloud Platform **without** sending scan results to Semgrep, use:
```

File: docs/getting-started/cli.md [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+semgrep ci --oss-only
+```
+
+:::info
+See [Semgrep Pro versus Semgrep OSS](/semgrep-pro-vs-oss) for information on the differences between Semgrep Pro and OSS Engines.
+:::
+
+## Scan using specific Semgrep Products
+
+When you run `semgrep ci`, you scan your project with Semgrep Code, Semgrep Supply Chain, and Semgrep Secrets. To scan your project with just one product, run:
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:135]:
```suggestion
When you run `semgrep ci`, you scan your project with any project that is enabled in the Cloud Platform. To scan your project with just one product, run:
```

File: docs/getting-started/cli.md [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+# scan with Semgrep Supply Chain
+semgrep ci --supply-chain
+
+# scan with Semgrep Secrets
+semgrep ci --secrets
+```
+
+## Extend timeout thresholds
+
+Depending on the file sizes in your project, you may need to increase the timeout threshold so that Semgrep doesn't time out before the scan completes. You can control this value using the `--timeout-threshold` flag, which refers to the maximum amount of time Semgrep spends scanning a single file. The default value is 30 seconds. Semgrep attempts to scan each file with this timeout value three times.
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:150]:
```suggestion
Depending on the file sizes in your project, you may need to increase the timeout threshold so that Semgrep doesn't time out before the scan completes. You can control this value using the `--timeout` flag, which refers to the maximum amount of time Semgrep spends scanning a single file. The default value is 5 seconds. Semgrep attempts to scan each file with this timeout value three times.
```
The time length is controlled with --timeout (default is 5 seconds in recent versions), and the number of times (default 3) is controlled with --timeout-threshold, so you could also add information on that after the last sentence.

File: docs/getting-started/cli.md [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+## Improve performance for large codebases
+
+You can set the number of subprocesses Semgrep uses to run checks in parallel:
+
+```console
+semgrep scan -j NUMBER_OF_SUBPROCESSES
+```
+
+By default, Semgrep defaults to the number of cores detected on the system on which Semgrep is running, but `-j = 1` if you're passing in `--pro`. For additional information, see [Parallelization](/kb/semgrep-code/scan-engine-kill)
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:165]: the link makes sense here :)

File: docs/getting-started/cli.md [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+To set the logging level for a scan, include the flag when scanning your project:
+
+```console
+# run a scan and get debug logs
+semgrep ci --debug
+```
+
+## Exit codes
+
+The CLI commands `semgrep ci` and `semgrep scan` finish with exit code `0` as long as the scan completes, regardless of whether there were findings. To finish with exit code `1` when there are findings, pass in the `--error` flag.
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:188]: This is not true. If you've configured blocking rules in SCP, `semgrep ci` will exit 1.

File: src/components/procedure/_login-activate.mdx [hunk 25]
```diff
@@ -0,0 +1,7 @@
+1. Log into your Semgrep account. Running this command launches a browser window, but you can also use the link that's returned in the CLI to proceed:
```
Comment (armchairlinguist) [inline src/components/procedure/_login-activate.mdx:1]:
```suggestion
1. Log in to your Semgrep account. Running this command launches a browser window, but you can also use the link that's returned in the CLI to proceed:
```

File: docs/getting-started/cli.md (hunk 17 above)
Comment (khorne3) [reply docs/getting-started/cli.md:39]: I don't mind removing, but it already warns users of this in the CLI; I thought it might be helpful to get ahead of the not-logged-in error?

File: docs/getting-started/cli.md [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+To set the logging level for a scan, include the flag when scanning your project:
+
+```console
+# run a scan and get debug logs
+semgrep ci --debug
+```
+
+## Exit codes
+
+The CLI commands `semgrep ci` and `semgrep scan` finish with exit code `0` as long as the scan completes, regardless of whether there were findings. To finish with exit code `1` when there are findings:
```
Comment (khorne3) [inline docs/getting-started/cli.md:193]: @armchairlinguist Do you mind taking a look at this section again? Not completely sure I explained it correctly (I also added an exit codes section to the OSS doc)

Comment (armchairlinguist) [reply docs/getting-started/cli.md:193]: Yeah, this isn't quite right yet. The first two are correct, but the the last item is incorrect. You only use `--no-suppress-errors` if you also want internal errors not to be suppressed. To get exit code 1, you only need rules in blocking mode that have findings in the scan.

File: docs/getting-started/cli.md [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+
+* [Configure blocking rules](/semgrep-code/policies/#blocking-a-pr-or-mr-through-rule-modes)
+* Pass in the `--error` flag when running `semgrep scan`.
+
+When you run `semgrep ci`, you can pass in the `--no-suppress-errors` if you don't want internal errors suppressed.
```
Comment (armchairlinguist) [inline docs/getting-started/cli.md:198]: Recommendation: link to the exit codes doc here also, to clarify what kinds of errors are internal. Otherwise looks good though! I only looked at the exit codes stuff this time around.

File: docs/getting-started/cli.md (hunk 17 above)
Comment (s-santillan) [reply docs/getting-started/cli.md:36]: Yeah the product already enforces it. We'd just be matching the current product behavior. Katie noticed it first; me not noticing this ASAP and realizing how much it impacts the docs is something I have to work on.

Comment (armchairlinguist) [reply docs/getting-started/cli.md:36]: @s-santillan Based on the latest thread, my understanding is that it doesn't enforce it yet (although it is something that may happen in the future). What it enforces now is that you can't both be logged in/provide a token, and use --config. If you're not logged in and use --config, it still works.

Comment (s-santillan) [reply docs/getting-started/cli.md:36]: Got it. This isn't in live yet, so I'll make the correction in my follow up PR (https://github.com/semgrep/semgrep-docs/pull/1350). cc @khorne3 (And yes I will fix the sidebars.)
